File: packages/csharp/src/StackFrame.cs
namespace Sererr;

/// <summary>
/// A single frame in a captured stack trace.
///
/// Field-compatible with the <c>sererr.v1.StackFrame</c> proto. All
/// fields use proto3 "zero value = unknown" semantics — defaults are
/// empty strings, empty lists, line=0, in_app=false.
/// </summary>
/// <remarks>
/// Property names use PascalCase per .NET convention; they map to the
/// proto's snake_case fields via the proto adapter.
/// </remarks>
public sealed record StackFrame
{
    /// <summary>Demangled function/method name.</summary>
    public string Function { get; init; } = "";

    /// <summary>Containing module / class / namespace.</summary>
    public string Module { get; init; } = "";

    /// <summary>Native library / crate / package name.</summary>
    public string Package { get; init; } = "";

    /// <summary>Source file (Sentry: <c>filename</c>).</summary>
    public string File { get; init; } = "";

    /// <summary>Absolute path on the build machine.</summary>
    public string AbsPath { get; init; } = "";

    /// <summary>1-based line number; 0 = unknown.</summary>
    public uint Line { get; init; }

    /// <summary>The exact source line at <see cref="Line"/>.</summary>
    public string ContextLine { get; init; } = "";

    /// <summary>Source lines preceding <see cref="Line"/>.</summary>
    public IReadOnlyList<string> PreContext { get; init; } = Array.Empty<string>();

    /// <summary>Source lines following <see cref="Line"/>.</summary>
    public IReadOnlyList<string> PostContext { get; init; } = Array.Empty<string>();

    /// <summary>Deep link to a source viewer for this frame.</summary>
    public string SourceLink { get; init; } = "";

    /// <summary>Hint to UIs: false for framework/stdlib frames.</summary>
    public bool InApp { get; init; }

    // Records' built-in Equals walks reference equality on collection
    // properties, which is wrong for IReadOnlyList. Override to use
    // sequence equality so two frames with equal field values compare
    // equal regardless of which list instance backs them.
    public bool Equals(StackFrame? other)
    {
        if (ReferenceEquals(this, other)) return true;
        if (other is null) return false;
        return Function == other.Function
            && Module == other.Module
            && Package == other.Package
            && File == other.File
            && AbsPath == other.AbsPath
            && Line == other.Line
            && ContextLine == other.ContextLine
            && PreContext.SequenceEqual(other.PreContext)
            && PostContext.SequenceEqual(other.PostContext)
            && SourceLink == other.SourceLink
            && InApp == other.InApp;
    }

    public override int GetHashCode()
    {
        var hc = new HashCode();
        hc.Add(Function);
        hc.Add(Module);
        hc.Add(Package);
        hc.Add(File);
        hc.Add(AbsPath);
        hc.Add(Line);
        hc.Add(ContextLine);
        foreach (var s in PreContext) hc.Add(s);
        foreach (var s in PostContext) hc.Add(s);
        hc.Add(SourceLink);
        hc.Add(InApp);
        return hc.ToHashCode();
    }
}

File: packages/java/src/main/java/fyi/sererr/StackFrame.java
package fyi.sererr;

import java.util.List;
import java.util.Objects;

/**
 * A single frame in a captured stack trace.
 *
 * <p>Field-compatible with the {@code sererr.v1.StackFrame} proto
 * definition. All fields use proto3 "zero value = unknown" semantics:
 * the empty string and the {@code 0} number mean "not populated."
 *
 * <p>Frames are ordered most-recent-call-first within a captured error,
 * matching the Sentry convention and Python's {@code traceback.format_exception}.
 *
 * <p>Records are immutable — to populate source-context fields after
 * capture, use {@link SourceContext#populateContext(StackFrame, SourceProvider, int)}
 * which returns a new frame with the populated fields.
 */
public record StackFrame(
        String function,
        String module,
        String pkg,
        String file,
        String absPath,
        int line,
        String contextLine,
        List<String> preContext,
        List<String> postContext,
        String sourceLink,
        boolean inApp
) {
    /** All-defaults constructor used for proto3 "zero value" semantics. */
    public static StackFrame empty() {
        return new StackFrame("", "", "", "", "", 0, "", List.of(), List.of(), "", false);
    }

    public StackFrame {
        Objects.requireNonNull(function, "function");
        Objects.requireNonNull(module, "module");
        Objects.requireNonNull(pkg, "pkg");
        Objects.requireNonNull(file, "file");
        Objects.requireNonNull(absPath, "absPath");
        Objects.requireNonNull(contextLine, "contextLine");
        Objects.requireNonNull(preContext, "preContext");
        Objects.requireNonNull(postContext, "postContext");
        Objects.requireNonNull(sourceLink, "sourceLink");
        preContext = List.copyOf(preContext);
        postContext = List.copyOf(postContext);
    }

    /** Builder helper for incrementally constructing frames in tests and adapters. */
    public static Builder builder() {
        return new Builder();
    }

    /** Mutable builder; not thread-safe. Call {@link #build()} to produce an immutable frame. */
    public static final class Builder {
        private String function = "";
        private String module = "";
        private String pkg = "";
        private String file = "";
        private String absPath = "";
        private int line = 0;
        private String contextLine = "";
        private List<String> preContext = List.of();
        private List<String> postContext = List.of();
        private String sourceLink = "";
        private boolean inApp = false;

        public Builder function(String v) { this.function = v; return this; }
        public Builder module(String v) { this.module = v; return this; }
        public Builder pkg(String v) { this.pkg = v; return this; }
        public Builder file(String v) { this.file = v; return this; }
        public Builder absPath(String v) { this.absPath = v; return this; }
        public Builder line(int v) { this.line = v; return this; }
        public Builder contextLine(String v) { this.contextLine = v; return this; }
        public Builder preContext(List<String> v) { this.preContext = v; return this; }
        public Builder postContext(List<String> v) { this.postContext = v; return this; }
        public Builder sourceLink(String v) { this.sourceLink = v; return this; }
        public Builder inApp(boolean v) { this.inApp = v; return this; }

        public StackFrame build() {
            return new StackFrame(function, module, pkg, file, absPath, line,
                    contextLine, preContext, postContext, sourceLink, inApp);
        }
    }
}
